### PR TITLE
route: add check for scheme headers during filter processing

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong <loong.dai@intel.com>

Commit Message: One littite part of #14961
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

/cc @asraa please check and review in case of any misunderstanding.

### CODE REVIEW

**Comment by mattklein123** [comment]
I will assign this over to @alyssawilk as I'm not sure what the plan is for this type of checking, but a few comments:
1) I don't think an error log is the right thing to do here. We should have a stat, a log that doesn't log very often, an ENVOY_BUG, or something else.
2) We need a test case for this.

Thank you.

/wait

**Comment by daixiang0** [comment]
@mattklein123 thanks for your review, I would add a test and wait input @alyssawilk.

File: source/common/router/router.cc [hunk 1]
```diff
@@ -673,7 +673,10 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
       headers, callbacks_->streamInfo().downstreamAddressProvider().sslConnection() != nullptr);
 
   // Ensure an http transport scheme is selected before continuing with decoding.
-  ASSERT(headers.Scheme());
+  if (!headers.Scheme()) {
```
Comment (alyssawilk) [inline source/common/router/router.cc:676]: Honestly I think we can leave this assert as-is or simply remove it it's basically validating that setUpstreamScheme sets a scheme, which it deterministically does.

Comment (mattklein123) [reply source/common/router/router.cc:676]: OK then I would probably just leave it and close this PR?

Comment (daixiang0) [reply source/common/router/router.cc:676]: @alyssawilk could you help check #14961, if all are like this, maybe we can close that as well.

Comment (daixiang0) [reply source/common/router/router.cc:676]: kindly ping @alyssawilk

Comment (alyssawilk) [reply source/common/router/router.cc:676]: yeah I mean I'm generally a fan of asserts so I'd be inclined to close off that issue, sorry

**Pull request closed without merging**